### PR TITLE
feat: Flip

### PR DESCRIPTION
**Pull request by wadjih-bencheikh18**
closes : https://github.com/image-js/image-js-typescript/issues/56

### CODE REVIEW

File: src/filters/flipX.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ */
+export default function flipX(image: IJS): IJS {
+  checkProcessable(image, 'flipX', {
+    bitDepth: [8, 16],
+  });
+  for (let i = 0; i < image.height; i++) {
+    for (let j = 0; j < Math.floor(image.width / 2); j++) {
+      const colCurrent = j;
+      const colOpposite = image.width - j - 1;
+      const tmp = image.getPixel(i, colCurrent);
```
Comment (lpatiny) [inline src/filters/flipX.ts:18]: Please add a loop on the channels. getPixel would create an array that would have to be garbadcollected and this is time consuming.

We try in image-js to have correct performance

File: src/filters/flip.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  copyData(image, newImage);
+  if (image.alpha) {
+    copyAlpha(image, newImage);
+  }
+  if (axis === 'horizontal') {
+    return flipX(newImage);
+  } else if (axis === 'vertical') {
+    return flipY(newImage);
+  } else {
+    return flipY(flipX(newImage));
```
Comment (targos) [inline src/filters/flip.ts:42]: Can you add a couple tests for this case?

File: src/filters/flip.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * @param options - Flip options.
+ * @returns - The processed image.
+ */
+export function flip(image: IJS, options: FlipOptions = {}): IJS {
+  const { axis = 'horizontal' } = options;
+  let newImage = getOutputImage(image, options);
+  copyData(image, newImage);
+  if (image.alpha) {
+    copyAlpha(image, newImage);
+  }
```
Comment (targos) [inline src/filters/flip.ts:36]: You can use the "clone: true" option of `getOutputImage` instead of this

File: src/filters/flipY.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * @returns The processed image.
+ */
+export default function flipY(image: IJS): IJS {
+  checkProcessable(image, 'flipY', {
+    bitDepth: [8, 16],
+  });
+
+  for (let i = 0; i < Math.floor(image.height / 2); i++) {
+    for (let j = 0; j < image.width; j++) {
+      const ligneCurrent = i;
```
Comment (lpatiny) [inline src/filters/flipY.ts:17]: Please use currentRow and oppositeRow

File: src/filters/flipX.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * @param image - Image to process.
+ * @returns The processed image.
+ */
+export default function flipX(image: IJS): IJS {
+  checkProcessable(image, 'flipX', {
+    bitDepth: [8, 16],
+  });
+  for (let i = 0; i < image.height; i++) {
+    for (let j = 0; j < Math.floor(image.width / 2); j++) {
+      const colCurrent = j;
```
Comment (lpatiny) [inline src/filters/flipX.ts:16]: Please use currentCol and oppositeCol

**Review by lpatiny — COMMENTED** [review]
Many thanks for this excellent contribution !

I did some small comments.

File: src/filters/flip.ts (hunk 2 above)
Comment (wadjih-bencheikh18) [reply src/filters/flip.ts:42]: Yeah no problem

File: src/filters/flipX.ts (hunk 1 above)
Comment (wadjih-bencheikh18) [reply src/filters/flipX.ts]: fixed

File: src/filters/flip.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ * @param image - Image to process.
+ * @param options - Flip options.
+ * @returns - The processed image.
+ */
+export function flip(image: IJS, options: FlipOptions = {}): IJS {
+  const { axis = 'horizontal' } = options;
+  let newImage = getOutputImage(image, options, { clone: true });
+  if (image.alpha) {
+    copyAlpha(image, newImage);
+  }
```
Comment (targos) [inline src/filters/flip.ts:34]: I think alpha is also copied with `clone: true`

Comment (wadjih-bencheikh18) [reply src/filters/flip.ts]: fixed thnx

**Pull request merged**